src/main/java/net/minecraftforge/event/AddReloadListenerEvent.java

```diff
@@ -20,33 +19,90 @@
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.Executor;
 
+import org.jetbrains.annotations.ApiStatus;
+
 /**
  * The main ResourceManager is recreated on each reload, just after {@link ReloadableServerResources}'s creation.
+ * A copy of the listeners attached from vanilla can be obtained from {@link ReloadableServerResources#listeners()}
  *
  * The event is fired on each reload and lets modders add their own ReloadListeners, for server-side resources.
  * The event is fired on the {@link MinecraftForge#EVENT_BUS}
  */
 public class AddReloadListenerEvent extends Event
 {
+    @Deprecated(forRemoval = true, since = "1.19.2")
     private final List<PreparableReloadListener> listeners = new ArrayList<>();
+    private final List<PreparableReloadListener> returnList;
     private final ReloadableServerResources serverResources;
 
+    @ApiStatus.Internal
+    public AddReloadListenerEvent(ReloadableServerResources serverResources, List<PreparableReloadListener> returnList)
+    {
+        this.serverResources = serverResources;
+        this.returnList = returnList;
+    }
+
+    /**
+     * @deprecated this constructor will be removed as it lacks the proper context
+     */
+    @Deprecated(forRemoval = true, since = "1.19.2")
+    @ApiStatus.Internal
     public AddReloadListenerEvent(ReloadableServerResources serverResources)
     {
         this.serverResources = serverResources;
+        this.returnList = new ArrayList<>();
     }
 
    /**
     * @param listener the listener to add to the ResourceManager on reload
     */
     public void addListener(PreparableReloadListener listener)
     {
-       listeners.add(new WrappedStateAwareListener(listener));
+        var wrapped = new WrappedStateAwareListener(listener);
+        listeners.add(wrapped);
+        returnList.add(wrapped);
     }
 
+
+    /**
+     * This method inserts a {@link PreparableReloadListener} before a specified vanilla reload listener. It does not work for inserting listeners before modded reload listeners. Use event priorities for that functionality.
+     *
```

Comment:
and if I may quote the javadoc

Reply:
Implementing a half baked system is not sufficient.  It must work for any reload listener if it is to exist.

The solution proposed internally was requiring all reload listeners be named, such that insertions can reference the name rather than the object (this also permits soft dependency sorting).

Reply:
Right, well in this case then these things should be proposed up front. I tried to write a simple solution that works for what I need and is backwards-compatible, I can't predict what is and is not considered half-baked. I am and have always been happy to rewrite PRs for your needs, but I need to know what those are. Further, if my solution is too complex, proposing something that involves further wrapping seems like an even bigger maintenance burden (what about when reload listeners are added in future versions? will the patch movers remember to add a name for it?)

Reply:
> Right, well in this case then these things should be proposed up front.  ..... 

Which is why Issues and discord exist. To gather information and requirements before writing the code. 1/2 the issue is that we don't have enough information to answer the "Why?" question. 

> Further, if my solution is too complex, proposing something that involves further wrapping seems like an even bigger maintenance burden (what about when reload listeners are added in future versions? will the patch movers remember to add a name for it?)

Well, for cases like this we typically make the code explode if vanilla doesn't have the data that we need it to have. So there isn't a way to forget about it. Beyond this is also related to our 'context' object that was discussed before. 

>Inserting before other modded entries is not supported.

Thats the main issue right now, you add a function to insert relative to another listener. But no way to actually know what the listener is. So it only works when being relative to vanilla's listeners. The current implementation just silently fails.